contracts/finance/andromeda-set-amount-splitter/src/contract.rs

```diff
@@ -0,0 +1,315 @@
+
+    SPLITTER.save(deps.storage, &splitter)?;
+
+    let inst_resp = ADOContract::default().instantiate(
+        deps.storage,
+        env,
+        deps.api,
+        &deps.querier,
+        info,
+        BaseInstantiateMsg {
+            ado_type: CONTRACT_NAME.to_string(),
+            ado_version: CONTRACT_VERSION.to_string(),
+            kernel_address: msg.kernel_address.clone(),
+            owner: msg.owner.clone(),
+        },
+    )?;
+
+    msg.validate(deps.as_ref())?;
+
+    Ok(inst_resp)
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn reply(_deps: DepsMut, _env: Env, msg: Reply) -> Result<Response, ContractError> {
+    if msg.result.is_err() {
+        return Err(ContractError::Std(StdError::generic_err(
+            msg.result.unwrap_err(),
+        )));
+    }
+
+    Ok(Response::default())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response, ContractError> {
+    let ctx = ExecuteContext::new(deps, info, env);
+
+    match msg {
+        ExecuteMsg::AMPReceive(pkt) => {
+            ADOContract::default().execute_amp_receive(ctx, pkt, handle_execute)
+        }
+        _ => handle_execute(ctx, msg),
+    }
+}
+
+pub fn handle_execute(mut ctx: ExecuteContext, msg: ExecuteMsg) -> Result<Response, ContractError> {
+    let action_response = call_action(
+        &mut ctx.deps,
+        &ctx.info,
+        &ctx.env,
+        &ctx.amp_ctx,
+        msg.as_ref(),
+    )?;
+    let res = match msg {
+        ExecuteMsg::UpdateRecipients { recipients } => execute_update_recipients(ctx, recipients),
+        ExecuteMsg::UpdateLock { lock_time } => execute_update_lock(ctx, lock_time),
+        ExecuteMsg::Send {} => execute_send(ctx),
+        _ => ADOContract::default().execute(ctx, msg),
+    }?;
+    Ok(res
+        .add_submessages(action_response.messages)
+        .add_attributes(action_response.attributes)
+        .add_events(action_response.events))
+}
+
+fn execute_send(ctx: ExecuteContext) -> Result<Response, ContractError> {
+    let ExecuteContext { deps, info, .. } = ctx;
+
+    // Maximum of two coins allowed for now
+    ensure!(
+        info.funds.len().le(&2),
+        ContractError::ExceedsMaxAllowedCoins {}
+    );
+
+    for coin in info.funds.clone() {
+        ensure!(
+            !coin.amount.is_zero(),
+            ContractError::InvalidFunds {
+                msg: "Amount must be non-zero".to_string(),
+            }
+        );
+    }
+
+    let splitter = SPLITTER.load(deps.storage)?;
+
+    let mut msgs: Vec<SubMsg> = Vec::new();
+    let mut amp_funds: Vec<Coin> = Vec::new();
+
+    let mut pkt = AMPPkt::from_ctx(ctx.amp_ctx, ctx.env.contract.address.to_string());
+
+    // Iterate through the sent funds
+    for coin in info.funds {
+        let mut remainder_funds = coin.amount;
+        let denom = coin.denom;
+
```

Comment:
What if denom is duplicated? Not sure if that is a possible case but still fund will be locked inside the contract in that case as `remainder_funds` will be calculated for the last fund.

A possible case scenario can be like this. I send fund via another ADO and the ADO is sending a message containing duplicated coin. In that case, the amount received on the recipient side will be doubled and splitter contract will have some funds.